components/patient-data/rest/src/main/java/org/phenotips/data/rest/internal/DefaultPatientByLabeledExternalIdentifierResourceImpl.java

```diff
@@ -0,0 +1,437 @@
+     * Tries to create a patient with provided {@code label} and {@code id} and {@code json data}. Returns a
+     * {@link Response} with no content if successful, an error code otherwise.
+     *
+     * @param label the name of the label
+     * @param id the id value for the label; will be overwritten if a label/id pair is specified in {@code jsonInput}
+     * @param jsonInput the incoming json from the request, used to update the created patient record
+     * @return a {@link Response} with no content if successful, an error code otherwise
+     */
+    private Response returnCreatePatientResponse(final String label, final String id, final JSONObject jsonInput)
+    {
+        this.logger.debug("Creating patient record with label [{}] and corresponding external ID [{}]", label, id);
+        try {
+            final User currentUser = this.users.getCurrentUser();
+            if (!this.access.hasAccess(currentUser, Right.EDIT,
+                this.currentResolver.resolve(Patient.DEFAULT_DATA_SPACE, EntityType.SPACE))) {
+                this.logger.error("Edit access denied to user [{}].", currentUser);
+                return Response.status(Response.Status.FORBIDDEN).build();
+            }
+            final Patient patient = this.repository.create();
+            return returnUpdatePatientResponse(patient, label, id, jsonInput, true, PatientWritePolicy.UPDATE);
+        } catch (final Exception ex) {
+            this.logger.error("Failed to create patient with label [{}] and corresponding external ID: [{}] from "
+                              + "JSON: {}.", label, id, jsonInput, ex);
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * Updates the {@code patient} with provided {@code jsonInput JSON data}. If a patient was created
+     * {@code wasCreated} and labeled identifiers {@code KEY_LABELED_EIDS} is not specified in the {@code jsonInput},
+     * then the {@code label} and {@code id} provided in the request path parameters are used to set the labeled
+     * identifiers in the patient.
+     *
+     * @param patient the {@link Patient} to update
+     * @param label the name of the label
+     * @param id the id value for the label
+     * @param jsonInput the patient data used to update, as {@link JSONObject}
+     * @param wasCreated whether a patient was created for the update
+     * @param policy the policy according to which patient data should be written
+     * @return a {@link Response} with no content if successful, an error code otherwise
+     */
+    private Response returnUpdatePatientResponse(final Patient patient, final String label, final String id,
+        final JSONObject jsonInput, boolean wasCreated, PatientWritePolicy policy)
+    {
+        if (hasInternalIdConflict(jsonInput, patient)) {
+            return Response.status(Response.Status.CONFLICT).build();
+        }
+        if (wasCreated && !jsonInput.has(KEY_LABELED_EIDS)) {
```

Comment:
Specified label should always be put in the input JSON, even if other labeled identifiers are set. Remove the check for `jsonInput.has(KEY_LABELED_EIDS)`.